File: src/pages/LoginPage.ts
import { Page, Locator } from '@playwright/test';
import { findSelfHealingElement } from '../core/PlaywrightExtensions';

export class LoginPage {
  constructor(private page: Page) {}

  private async getUsernameField(): Promise<Locator> {
    return await findSelfHealingElement(this.page, '#UserName');
  }

  private async getPasswordField(): Promise<Locator> {
    return await findSelfHealingElement(this.page, '#Password');
  }

  private async getLoginButton(): Promise<Locator> {
    return await findSelfHealingElement(this.page, 'input[type="submit"]');
  }

  async login(username: string, password: string): Promise<void> {
    const usernameField = await this.getUsernameField();
    await usernameField.fill(username);

    const passwordField = await this.getPasswordField();
    await passwordField.fill(password);

    const loginButton = await this.getLoginButton();
    await loginButton.click();
  }
}

File: SeleniumSelfHealingTests/Pages/LoginPage.cs
using OpenQA.Selenium;
using SeleniumSelfHealingTests.Utilities.Extensions;

namespace SeleniumSelfHealingTests.Pages;

public class LoginPage
{
    public LoginPage(IWebDriver driver)
    {
        _driver = driver;
    }
    
    private readonly IWebDriver _driver;
    private Task<IWebElement> UsernameField => _driver.AiFindElement(By.Id("UserName"));
    private Task<IWebElement> PasswordField => _driver.AiFindElement(By.Id("Password"));
    private Task<IWebElement> LoginButton => _driver.AiFindElement(By.CssSelector("input[type='submit']"));
    
    public async Task Login(string userName, string password)
    {
       await UsernameField.SendKeys(userName);
       await PasswordField.SendKeys(password);
       await LoginButton.Click();
    }
  
}
